modules/core/src/effects/shadow-effect.js

```diff
@@ -0,0 +1,105 @@
+import Effect from '../lib/effect';
+import ShadowPass from '../passes/shadow-pass';
+import {Matrix4, Vector3} from 'math.gl';
+import {Texture2D} from '@luma.gl/core';
+
+export default class ShadowEffect extends Effect {
+  constructor(props) {
+    super(props);
+    const {shadowColor = [2, 0, 5, 200], lights = []} = props;
+    this.shadowPasses = [];
+    this.shadowColor = shadowColor;
+    this.lights = lights;
+    this.lightMatrices = [];
+    this.dummyShadowMaps = [];
+
+    this._createLightMatrix();
+  }
+
+  prepare(gl, {layers, viewports, onViewportActive, views, effects, pixelRatio}) {
+    if (this.shadowPasses.length === 0) {
+      this._createShadowPasses(gl, pixelRatio);
+    }
+
+    if (this.dummyShadowMaps.length === 0) {
+      this._createDummyShadowMaps(gl);
+    }
+
+    const shadowMaps = [];
+
+    for (let i = 0; i < this.shadowPasses.length; i++) {
+      const shadowPass = this.shadowPasses[i];
+      shadowPass.render({
+        layers,
+        viewports,
+        onViewportActive,
+        views,
+        effects,
+        effectProps: {
+          shadow_lightId: i,
+          dummyShadowMaps: this.dummyShadowMaps,
+          shadow_viewProjectionMatrices: this.lightMatrices
+        }
+      });
+      shadowMaps.push(shadowPass.shadowMap);
+    }
+
+    return {
+      shadowMaps,
+      dummyShadowMaps: this.dummyShadowMaps,
+      shadow_lightId: 0,
+      shadowColor: this.shadowColor,
+      shadow_viewProjectionMatrices: this.lightMatrices
+    };
+  }
+
+  cleanup() {
+    for (const shadowPass of this.shadowPasses) {
+      shadowPass.delete();
+    }
+    this.shadowPasses.length = 0;
+
+    for (const dummyShadowMap of this.dummyShadowMaps) {
+      dummyShadowMap.delete();
+    }
+    this.dummyShadowMaps.length = 0;
+  }
+
+  _createLightMatrix() {
+    const projectionMatrix = new Matrix4().ortho({
```

Comment:
Why isn't this a constant declared once at the top of the file?

Reply:
do you mean the projectionMatrix? we can create one in the top of file, initially I just want it to stay in its local scope cause we only create it one time when effect is constructed.